src/internal.cc

```diff
@@ -162,18 +162,36 @@ internalHostname(void)
     return host;
 }
 
-int
-internalHostnameIs(const char *arg)
+bool
+internalUriTargetingListeningPort(const AnyP::Uri &url, const AnyP::PortCfg &listeningPort)
 {
-    wordlist *w;
-
-    if (0 == strcmp(arg, internalHostname()))
-        return 1;
-
-    for (w = Config.hostnameAliases; w; w = w->next)
-        if (0 == strcmp(arg, w->key))
-            return 1;
+    // different port numbers is a fast NO
+    if (url.port() != listeningPort.s.port())
+        return false;
+
+    // accept raw-IP URL matching a port configured with explicit IP:port settings
+    if (url.hostIsNumeric() && !listeningPort.s.isAnyAddr() && url.hostIP() == listeningPort.s)
+        return true;
```

Comment:
Should not we reject specific IP mismatches?

```suggestion
    if (url.hostIsNumeric() && !listeningPort.s.isAnyAddr())
        return url.hostIP() == listeningPort.s;
```

If we want to preserve the official code logic with regard to an internalHostname() and hostname_aliases match (for hostIsNumeric() cases), then we would _also_ need to move this check further down, after those checks. The answer to that question may depend on whether we want to support IP addresses as internalHostnames and hostname_aliases. 

Another potential related concern here is whether it is possible to specify a listening port IP mask like 127.0.0.255. I cannot check that quickly right now, but if it is possible, then the equality comparison for IPs is wrong.

Reply:
Thanks. Yes we should.